quint/src/quintVerifier.ts

```diff
@@ -311,11 +315,141 @@ async function retryWithTimeout<T>(f: () => Promise<T>): Promise<VerifyResult<T>
+  fs.mkdirSync(unpackPath, { recursive: true })
+
+  return fetch(url).then(
+    res =>
+      new Promise((resolve, _) => {
+        const untar = tar.extract(unpackPath)
+        res.body.pipe(gunzip()).pipe(untar)
+        untar.on('close', () => resolve(right(unpackPath)))
+        untar.on('error', error => resolve(err(`Error unpacking .tgz: ${error}`)))
+      }),
+    error => err(`Error fetching ${url}: ${error}`)
+  )
+}
+
+/**
+ * Fetch the latest Apalache release pinned by `APALACHE_VERSION_TAG` from Github.
+ *
+ * @returns A promise resolving to:
+ *    - a `right<string>` equal to the path the Apalache dist was unpacked to,
+ *    - a `left<VerifyError>` indicating an error.
+ */
+async function fetchApalache(): Promise<VerifyResult<string>> {
+  // Fetch Github releases, find latest that satisfies `APALACHE_VERSION_TAG`
+  return octokitRequest('GET /repos/informalsystems/apalache/releases').then(
+    async resp => {
+      const releases = resp.data.map((element: { tag_name: string }) => element.tag_name)
+      const latestTaggedVersion = semver.maxSatisfying(releases, APALACHE_VERSION_TAG)
+      // Filter release response to get dist archive asset URL
+      const taggedReleases = resp.data.filter(
+        (element: { tag_name: string }) => element.tag_name == latestTaggedVersion
+      )
+      const tgzAssets = taggedReleases[0].assets.filter((asset: { name: string }) => asset.name == APALACHE_TGZ)
+      const downloadUrl = tgzAssets[0].browser_download_url
+      // Check if we have already downloaded this release
+      const unpackPath = path.join(os.homedir(), '.quint', `apalache-dist-${latestTaggedVersion}`)
+      const apalacheBinary = path.join(unpackPath, 'apalache', 'bin', 'apalache-mc')
+      if (fs.existsSync(apalacheBinary)) {
+        // Use existing download
+        console.log(`Using existing Apalache distribution in ${unpackPath}`)
+        return right(unpackPath)
+      } else {
+        // Download Apalache dist
+        console.log(`Downloading ${downloadUrl}`)
+        return await fetchAndUnpack(downloadUrl, unpackPath)
+      }
+    },
+    error => err(`Fetching Apalache releases failed: ${error}`)
+  )
+}
+
+/**
+ * Connect to an already running Apalache server, or – if unsuccessful – fetch
+ * Apalache, spawn it, and connect to it.
+ *
+ * If an Apalache server is spawned, the child process exits when the parent process (i.e., this process) terminates.
+ *
+ * @returns A promise resolving to:
+ *    - a `right<Apalache>` equal to the path the Apalache dist was unpacked to,
+ *    - a `left<VerifyError>` indicating an error.
+ */
+async function connect(): Promise<VerifyResult<Apalache>> {
+  // Try to connect to Shai, and try to ping it
+  const connectionResult = await tryConnect()
+  // We managed to connect, simply return this connection
+  if (connectionResult.isRight()) {
+    return connectionResult
+  }
```

Comment:
As is, this means waiting 5 seconds every startup (if the server isn't running). We can fix that by adding a way to try getting a connection without retrying and waiting.

Reply:
Uh, you're right. I assumed that once we obtain an `AsyncCmdExecutor` and the server is already up, the ping would succeed immediately. I will change this.

Reply:
Reworked the retry logic in [1a6a1d5](https://github.com/informalsystems/quint/pull/1115/commits/1a6a1d563ff9c1ca7d92cb5cf9c10393f2f57a2b)